### ISSUE TITLE
False negative for protocols with isinstance check

### ISSUE THREAD

**Issue by luwqz1** [microsoft/pyright#10276]
**Describe the bug**
I started investigating issue #10270 and encountered this bug while checking whether the value `x` conforms to the `Foo` proto. The type of the value `x` is a union of `Foo` and an inheritable type. The example includes two cases: in the first, the type is a subclass of str, and in the second, it is the str class itself.

**Code or Screenshots**

```python
from typing import Protocol, runtime_checkable, reveal_type


@runtime_checkable
class Foo(Protocol):
   def foo(self) -> int: ...


class MyStr(str):
    def foo(self) -> str:
        return self


def func1(x: str | Foo) -> None:
    if isinstance(x, Foo):  # false negative
        reveal_type(x.foo().bit_count())


def func2(x: MyStr | Foo) -> None:
    if isinstance(x, Foo):  # (reportGeneralTypeIssues) as expected
        reveal_type(x.foo().bit_count())


func1(MyStr())
func2(MyStr())
```

[pyright playground](https://pyright-play.net/?pythonVersion=3.13&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAK4MYAxmADYA0UIArijFgKYD6pAFs6QNYCGAIwrMaIZgDdmfCq3gJmAKCUABeoxbsuvQcIWkKfAM6GoAMTBgAFETAlyFAJQAuBVDcATZsCjALlw8wUwA5QALQAfJiMTlAAdPFKegbGUACycADKMCD%2B2c6ublCe3r5WAUEhEVCG2S6F9bTMMHQgKNWBwInFPgykAIyWAB4xNSBQAD5mFpWRAHJgKMx1hUjeSIaoNXwopMxDNOZgzm4AxD7SAVCLaHxMkgUNjZLSsoh7g7Gllg6xAkgw7DADBg3wcXS8PR2ACYhjF0lkxpNDjMoPNFss3KtMBsUFsdu8DtMYlAzpYxDg8ABxZiLEDSAAqbwAksY6MxDCEjFBmIN5KQYMx3A8GmJnjI5O9Pn4fn8AeRgaDEsBegN4dlFcroZY1TkHGCgA)

**Comment by erictraut** [comment]
Pyright is working correctly here and in conformance with the typing spec. For details, refer to the rules in [this section](https://typing.python.org/en/latest/spec/protocol.html#runtime-checkable-decorator-and-narrowing-types-by-isinstance). In the first example, there is no "unsafe overlap" between the protocol `Foo` and the first argument passed to `isinstance`. In the second example, there is "unsafe overlap".

**Comment by luwqz1** [comment]
Thanks for your answer. I have a question: in the case of `func1`, isn't this actually is no "unsafe overlap"? I’m asking because it's possible to pass a subclass of str there, which may have the same attributes as the protocol but return different types at runtime, as shown in the example. Is such behavior acceptable for a static type checker?

I initially assumed that protocols should only be used with types marked with the `final` decorator, indicating that subclasses are not allowed, but it seems this all depends on the typing spec, or perhaps I don't fully understand these cases. I’d appreciate it if you could clarify these points or share any sources where I can read more about this for better understanding. Thank you!

**Comment by erictraut** [comment]
The term "unsafe overlap" comes from the typing spec and has a specific meaning in a specific context. The `func1` example isn't an "unsafe overlap" as the typing spec defines it.

The typing spec provides little or no guidance when it comes to type narrowing behaviors. Type checkers like mypy and pyright strive to strike a balance between strictness and practicality. What you're saying about subclasses is true, but this is a case where being really strict would make the functionality difficult to use for most Python code. The goal of pyright isn't to find every possible bug in every piece of code. The goal is to find common problems in typical Python code. If a type checker is overly strict, its adoption will be low. 

In any case, pyright is working as intended here. This isn't a bug.

**Comment by luwqz1** [comment]
Thank you very much for your answer! I'm glad that the pyright works as intended.
